plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  AZURE_DEVOPS_BUILD_DEFINITION_ANNOTATION,
+  AZURE_DEVOPS_HOST_ORG_ANNOTATION,
+  AZURE_DEVOPS_PROJECT_ANNOTATION,
+  AZURE_DEVOPS_REPO_ANNOTATION,
+} from '../constants';
+
+import { Entity } from '@backstage/catalog-model';
+
+export function getAnnotationValuesFromEntity(entity: Entity): {
+  project: string;
+  repo?: string;
+  definition?: string;
+  host?: string;
+  org?: string;
+} {
+  const { host, org } = getHostOrg(entity.metadata.annotations);
+
+  const projectRepoValues = getProjectRepo(entity.metadata.annotations);
+  if (projectRepoValues.project && projectRepoValues.repo) {
+    return {
+      project: projectRepoValues.project,
+      repo: projectRepoValues.repo,
+      host,
+      org,
+    };
+  }
+
+  const project =
+    entity.metadata.annotations?.[AZURE_DEVOPS_PROJECT_ANNOTATION];
+  if (!project) {
+    throw new Error(
+      `Value for annotation "${AZURE_DEVOPS_PROJECT_ANNOTATION}" was not found`,
+    );
+  }
+
+  const definition =
+    entity.metadata.annotations?.[AZURE_DEVOPS_BUILD_DEFINITION_ANNOTATION];
+  if (!definition) {
+    throw new Error(
+      `Value for annotation "${AZURE_DEVOPS_BUILD_DEFINITION_ANNOTATION}" was not found`,
+    );
+  }
+  return { project, definition, host, org };
+}
+
+function getProjectRepo(annotations?: Record<string, string>): {
+  project?: string;
+  repo?: string;
+} {
+  const annotation = annotations?.[AZURE_DEVOPS_REPO_ANNOTATION];
+  if (!annotation) {
+    return { project: undefined, repo: undefined };
+  }
+
+  if (annotation.includes('/')) {
+    const [project, repo] = annotation.split('/');
```

Comment:
this silently rejects anything past a second slash - no big deal, just mentioning that it's not super strict on input

Reply:
This one we are safe, you can not use `/` in project or repo names 👍

Reply:
oh but you know, accidents etc

Reply:
Right, will think on this as well

Reply:
https://github.com/backstage/backstage/pull/21797 resolves this too